core/distributed/vector.cpp

```diff
@@ -0,0 +1,394 @@
+}
+
+
+template <typename ValueType>
+std::unique_ptr<typename Vector<ValueType>::complex_type>
+Vector<ValueType>::make_complex() const
+{
+    auto result = complex_type::create(
+        this->get_executor(), this->get_communicator(), this->get_size(),
+        this->get_const_local()->get_size(),
+        this->get_const_local()->get_stride());
+    this->make_complex(result.get());
+    return result;
+}
+
+
+template <typename ValueType>
+void Vector<ValueType>::make_complex(Vector::complex_type* result) const
+{
+    this->get_const_local()->make_complex(result->get_local());
+}
+
+
+template <typename ValueType>
+std::unique_ptr<typename Vector<ValueType>::real_type>
+Vector<ValueType>::get_real() const
+{
+    auto result =
+        real_type::create(this->get_executor(), this->get_communicator(),
+                          this->get_size(), this->get_const_local()->get_size(),
+                          this->get_const_local()->get_stride());
+    this->get_real(result.get());
+    return result;
+}
+
+
+template <typename ValueType>
+void Vector<ValueType>::get_real(Vector::real_type* result) const
+{
+    this->get_const_local()->get_real(result->get_local());
+}
+
+
+template <typename ValueType>
+std::unique_ptr<typename Vector<ValueType>::real_type>
+Vector<ValueType>::get_imag() const
+{
+    auto result =
+        real_type::create(this->get_executor(), this->get_communicator(),
+                          this->get_size(), this->get_const_local()->get_size(),
+                          this->get_const_local()->get_stride());
+    this->get_imag(result.get());
+    return result;
+}
+
+
+template <typename ValueType>
+void Vector<ValueType>::get_imag(Vector::real_type* result) const
+{
+    this->get_const_local()->get_imag(result->get_local());
+}
+
+
+template <typename ValueType>
+void Vector<ValueType>::scale(const LinOp* alpha)
+{
+    this->get_local()->scale(alpha);
+}
+
+
+template <typename ValueType>
+void Vector<ValueType>::inv_scale(const LinOp* alpha)
+{
+    this->get_local()->inv_scale(alpha);
+}
+
+
+template <typename ValueType>
+void Vector<ValueType>::add_scaled(const LinOp* alpha, const LinOp* b)
+{
+    auto dense_b = as<Vector<ValueType>>(b);
+    this->get_local()->add_scaled(alpha, dense_b->get_const_local());
+}
+
+
+template <typename ValueType>
+void Vector<ValueType>::sub_scaled(const LinOp* alpha, const LinOp* b)
+{
+    auto dense_b = as<Vector<ValueType>>(b);
+    this->get_local()->sub_scaled(alpha, dense_b->get_const_local());
+}
+
+
+template <typename ValueType>
+void Vector<ValueType>::compute_dot(const LinOp* b, LinOp* result) const
+{
+    auto exec = this->get_executor();
+    const auto comm = this->get_communicator();
+    auto dense_res =
+        make_temporary_clone(exec, as<matrix::Dense<ValueType>>(result));
+    this->get_const_local()->compute_dot(as<Vector>(b)->get_const_local(),
+                                         dense_res.get());
+    exec->synchronize();
```

Comment:
Is this necessary with non GPU-aware MPI?

Reply:
I dont think this is necessary even with GPU-aware MPI. All our kernel calls are currently launched on the default stream. So, everything is completely synchronous.

Reply:
As far as I understood a few discussions on OpenMPI's design in a related issue, you cannot configure which stream (if any) MPI uses to execute the transfers. Might even be DMA without involvement of the SMs? So I think this will always be necessary.